File: coding/python/programmers/level1/dart_game.py
def solution(dartResult):
    values = []  # 숫자 저장용
    operators = []  # 연산자 저장용
    idx = 0
    count = 0

    for i in range(0, len(dartResult)):
        if dartResult[i].isdigit():  # 숫자 여부
            if dartResult[i] == '0' and i > 0 and dartResult[i-1].isdigit():  # '1','0' 을 10 으로.
                values[-1] = 10
            else:
                values.append(int(dartResult[i]))
        else:
            operators.append(dartResult[i])  # 연산자

    while len(operators) > idx:
        if operators[idx] in ['S', 'D', 'T']:
            if operators[idx] == 'D':
                values[count] = (values[count] * values[count])  # 제곱
            elif operators[idx] == 'T':
                values[count] = (values[count] * values[count] * values[count])  # 세제곱

            count += 1
            idx += 1
        else:  # '*' 또는 '#'
            if operators[idx] == '*':  # 해당 점수를 2배로.
                values[count-1] = values[count-1] * 2
                if idx > 1:  # 바로 전에 얻은 점수도 2배로.
                    values[count-2] = values[count-2] * 2
            elif operators[idx] == '#':  # 해당 점수를 마이너스 처리.
                values[count-1] = values[count-1] * (-1)

            idx += 1

    return sum(values)


if __name__ == "__main__":
    print(solution("1S2D*3T"))  # 37
    print(solution("1D2S#10S"))  # 9

File: coding/swift/Problem-Solving/Problem-Solving/Programmers/Level1/DartGame.swift
import Foundation

func solution(_ dartResult:String) -> Int {
    let values = Array(dartResult)
    
    var lastNumIdx = 0
    var results = Array(repeating: 0, count: 3)
    
    for i in 0..<values.count {
        
        if !(values[i].isNumber) && values[i] != "*" && values[i] != "#" {
            var temp = 0
            
            if values[i-1] == "0" && i >= 2 && values[i-2] == "1" {
                temp = 10
            } else {
                temp = values[i-1].wholeNumberValue!
            }
            
            if values[i] == "S" {
                results[lastNumIdx] = temp
            } else if values[i] == "D" {
                results[lastNumIdx] = (temp * temp)
            } else if values[i] == "T" {
                results[lastNumIdx] = (temp * temp * temp)
            }
            
            lastNumIdx += 1
        } else {
            if values[i] == "*" {
                results[lastNumIdx-1] *= 2
                
                if lastNumIdx >= 2 {
                    results[lastNumIdx-2] *= 2
                }
            } else if values[i] == "#" {
                results[lastNumIdx-1] *= -1
            }
        }
    }
    
    return results.reduce(0) { $0 + $1 }
}

//print(solution("1S2D*3T"))  // 37
//print(solution("1D2S#10S"))  // 9
//print(solution("1D2S0T"))  // 3
//print(solution("1S*2T*3S"))  // 23
//print(solution("1D#2S*3S"))  // 5
//print(solution("1T2D3D#"))  // -4
//print(solution("1D2S3T*"))  // 59
